packages/x-components/src/x-modules/search-box/components/__tests__/search-input.spec.ts

```diff
@@ -176,6 +176,15 @@ describe('testing search input component', () => {
     }
   );
 
+  it('prevents the query from containing special characters.', () => {
+    const queries: string[] = ['wa>ter', '<water', 'w<>ater', 'wa>ter>', '<>water', 'water><'];
+    for (const query of queries) {
+      input.value = query;
+      mockedSearchInput.trigger('keydown');
+      expect(input.value).toEqual('water');
+    }
+  });
```

Comment:
Thanks for the review @javieri-empathy! So currently I have made the function to remove all forbidden characters if the user pastes a query and it is tested in the search-input.spec.ts but if I were to try and test the `preventDefault` is working properly to prevent the user from typing the forbidden characters, should I add another test file to the ../search-box/components/__tests__/ directory that uses cypress or in the x-components/tests/unit/ directory? Since I cannot use due-test-utils and cypress in the same folder. Could you please help me with this? Thanks!

Reply:
@wannieman98 we are live coding in Twitch just in case you are interested ;)

https://www.twitch.tv/empathyco

Reply:
Sure, we use Cypress for both e2e tests together with Cucumber, but also for some component tests that rely too much on the browser. I would test this behavior using the last one, Cypress Component Testing.

Our Cypress component tests are located under `x-components/tests/unit`. There you can see a few examples on how  we write this tests. What we do is create a `mountXXX` function that does the "setup" step for each test. This mount function can receive parameters if they are needed, and return a basic API so the tests remain DRY, but readable.

For example, for the search input we could write this mount function. Note that we have to do some weird things like importing `reflect-metadata` due to our old adapter, or resetting the XPlugin and the registered modules each time, to ensure that each use case runs on fresh instances:

```ts
import 'reflect-metadata';
import { mount } from '@cypress/vue';
import Vue from 'vue';
import { XPlugin } from '../../src/plugins/x-plugin';
import { BaseXBus } from '../../src/plugins/x-bus';
import { mockedAdapter } from '../../src/adapter/mocked-adapter';
import SearchInput from '../../src/x-modules/search-box/components/search-input.vue';
import { searchBoxXModule } from '../../src/x-modules/search-box';

function mountSearchInput(): MountSearchInputAPI {
  XPlugin.resetInstance();
  mount(
    {
      components: {
        SearchInput
      },
      template: `
        <SearchInput />
      `
    },
    {
      vue: Vue.extend({}),
      plugins: [
        [
          new XPlugin(new BaseXBus()),
          { adapter: mockedAdapter, initialXModules: [searchBoxXModule] }
        ]
      ]
    }
  );

  return {
    getSearchInput() {
      return cy.getByDataTest('search-input');
    }
  };
}
```

Then, Cypress uses a similar syntax to the jest one for defining use cases, but a different one to run assertions. To test that invalid characters are ignored you could simply add this.

```ts
describe('testing search input', () => {
  it('ignores invalid typed characters', () => {
    const { getSearchInput } = mountSearchInput();
    getSearchInput().type('le<g>o >star>').invoke('val').should('equal', 'lego star');
  });
});
```

I've been trying to also write a test for pasting text, but I've failed, none of the approaches I found (dispatching events, typing, modifying the input value manually and triggering a fake event) works for this case. If you find anything I would be more than happy to check it.